[DirectX] Start documenting DXIL Resource handling

bogner opened the pull request:
This adds a new document about DXIL Resource Handling. I've attempted to describe here how we intend to use TargetExtTypes to represent resources in LLVM IR and the various intrinsics we'll need to lower these through LLVM to DXIL.

For now this document is limited to the high level concepts and a few details on buffer types, and there are a number of TODOs in the document that we'll iterate on as we progress in the implementation.

bogner committed:
[DirectX] Start documenting DXIL Resource handling

This adds a new document about DXIL Resource Handling. I've attempted
to describe here how we intend to use TargetExtTypes to represent
resources in LLVM IR and the various intrinsics we'll need to lower
these through LLVM to DXIL.

There are a number of TODOs in the document, but I think it makes
sense to publish as is and evolve it as we start implementing this.

python3kgae commented on llvm/docs/DirectX/DXILResources.rst:
```diff
@@ -0,0 +1,799 @@
+   binding space, lower bound in that space, and size of the binding, or (b) an
+   index into a dynamic resource heap.
+
+   In LLVM we represent binding information in the arguments of the
```
So clang codeGen will not create metadata which saves resource binding?
Because the binding information is in createHandleFromBinding?

bogner replied:
Yes. The information in the metadata in DXIL is represented in the create handle calls and target types.

python3kgae replied:
How does an array of resources work?
Always initialize the first element with createHandleFromBinding in clang codeGen to make sure the binding is saved?

bogner replied:
Yeah, arrays of resources will be handled similarly to how they're handled in dxc where the binding's size tells us the information we need.

damyanp commented on llvm/docs/DirectX/DXILResources.rst:
```diff
@@ -0,0 +1,799 @@
+
+`Binding information`
+   Information about where the resource comes from. This is either (a) a
+   binding space, lower bound in that space, and size of the binding, or (b) an
```
what is a "binding space"?

bogner replied:
It's the "space" in `register(t3,  space0)`. From https://learn.microsoft.com/en-us/windows/win32/direct3d12/resource-binding-in-hlsl:

> The space keyword specifies to which logical register space the declared variable is bound. If the space keyword is omitted, then the default space index of 0 is implicitly assigned to the range (so the tex2 range above resides in space0). register(t3, space0) will never conflict with register(t3, space1), nor with any array in another space that might include t3.

damyanp replied:
I've always called those "register spaces" (and it seems that's what the linked doc calls them as well).

damyanp replied:
If we want to use some different nomenclature for this here then I think it's probably worth explaining that.

bogner replied:
It's called "bind space" in https://github.com/microsoft/DirectXShaderCompiler/blob/main/docs/DXIL.rst#metadata-resource-records for whatever reason. I can update to use register space instead, as I do think that's more common.

damyanp commented on llvm/docs/DirectX/DXILResources.rst:
```diff
@@ -0,0 +1,799 @@
+
+   target("dx.Sampler", SamplerType)
+
+The "dx.Sampler" type is used to represent sampler state. The sampler type is
+an enum value from the DXIL ABI, and these appear in sampling operations as
```
> The sample type is an enum value from the DXIL ABI

Do we want to tie this so closely to the DXIL ABI?  Would an extra level of indirection here allow LLVM and DXIL to evolve independently?

damyanp commented on llvm/docs/DirectX/DXILResources.rst:
```diff
@@ -0,0 +1,799 @@
+
+The "dx.CBuffer" type is a constant buffer of the given size. Note that despite
+the name this is distinct from the buffer types, and can only be read using the
+"dx.CBufferLoad" and "dx.CBufferLoadLegacy" operations.
```
> "dx.CBufferLoad"

Didn't we just remove support for this from DXC? (I may be misunderstanding things)

bogner replied:
Not exactly, but this isn't the right place to discuss the legacy operation. I touch on it briefly in the "Constant Buffer Loads" section and I've updated this part to just mention `llvm.dx.cbufferLoad`

damyanp commented on llvm/docs/DirectX/DXILResources.rst:
```diff
@@ -0,0 +1,799 @@
+.. code-block:: llvm
+
+   target("dx.Texture1D", ElementType, IsWriteable, IsROV)
+   target("dx.Texture1D", ...)
```
Is it significant that `dx.Texture1D` is listed twice here?

bogner replied:
Typo - fixed.

damyanp commented on llvm/docs/DirectX/DXILResources.rst:
```diff
@@ -0,0 +1,799 @@
+various sample and texture load/store operations such that their parameters are
+appropriate to the type.
+
+.. list-table:: Buffer Fields
```
```suggestion
.. list-table:: Texture Fields
```

damyanp commented on llvm/docs/DirectX/DXILResources.rst:
```diff
@@ -0,0 +1,799 @@
+   * - FeedbackType
+     - Feedback type for a feedback texture.
+
+.. note:: TODO: Should dimensions be specified as a type?
+
+   We could probably get rid of the 1D/2D/etc distinctions...
```
> Should dimensions be specified as a type?

Seems that you made a good argument earlier for why not to do this above?

> so that we can overload the various sample and texture load/store operations such that their parameters are appropriate to the type.

bogner replied:
The TODO under "Texture Load and Store" covers what I was getting at here more thoroughly. I've removed this one.

bogner committed:
editorial fixes

bogner committed:
Use "register space" instead of "bind space"

python3kgae commented on llvm/docs/DirectX/DXILResources.rst:
```diff
@@ -0,0 +1,794 @@
+
+.. code-block:: llvm
+
+   target("dx.CBuffer", BufferSize)
```
Why not save the type like the dx.Buffer but a BufferSize?
For cbuffer we'll create a dummy struct to cache the layout, for ConstantBuffer<> case, the element type is already there.

bogner replied:
Removed the cbuffer stuff from this doc for now. There are a couple of trade offs on how to represent the type in a cbuffer, but just using an LLVM type doesn't seem optimal since the mapping between types and cbuffer layout is a bit complicated.

bogner committed:
two buffer types

bogner committed:
Update dxil buffer load/store docs

bogner committed:
cut DXILResources.rst down to just the buffer parts

A lot of this document is in flux, but it would be helpful to get the
parts we're already working on in tree. To facilitate that, cut out
everything that isn't about handles or buffers.

bogner commented on llvm/docs/DirectX/DXILResources.rst:
```diff
@@ -0,0 +1,794 @@
+The types involved in the store intrinsics match the load intrinsics.
+
+When lowering these to the DXIL operations we need to pay attention to the DXIL
+version and the type of data in the buffer. Post DXIL 1.2 structured and byte
```
Current version differentiates between raw/structured and typed loads, so this should all be fairly straightforward. Whether we lower raw/structured loads in one or two steps when targetting dxil 1.2 is mostly an implementation detail - we can do whatever makes more sense.

damyanp commented on clang/docs/HLSL/ResourceTypes.rst:
```diff
@@ -29,6 +29,7 @@ pointer of the template parameter type. The pointer is populated from a call to
 Resource types are annotated with the ``HLSLResource`` attribute, which drives
-code generation for resource binding metadata. The ``hlsl`` metadata nodes are
-transformed in the backend to the binding information expected by the target
-runtime.
+code generation into target extension types in IR. These types are target
+specific and differ between DXIL and SPIR-V generation, providing the necessary
+information for the targets to generate binding metadata for their respective
+target runtimes.
```
I think that this needs to be reconciled with #98419?

bogner replied:
Good point. I've updated this doc to talk about `__hlsl_resource_t` and to talk about the set of resource attributes more generically.

bogner committed:
Update resource types to discuss `__hlsl_resource_t`

bogner commented:
Just noticed that this was missing the change I had locally for updating the load and store operations to reflect the two different buffer types. I've pushed that in the latest update.

llvm-beanz commented on llvm/docs/DirectX/DXILResources.rst:
```diff
@@ -0,0 +1,425 @@
+``dx.op.createHandleFromHeap``, and ``dx.op.createHandleForLib``, but they fold
+the subsequent ``dx.op.annotateHandle`` operation in. Note that we don't have
+an analogue for `dx.op.createHandle`_, since ``dx.op.createHandleFromBinding``
+subsumes it.
```
Is the intent to translate `llvm.dx.handle.FromBinding` to `dx.op.createHandle` directly for older shader models?

bogner replied:
Yes.
